### PR TITLE
otelgrpc: Implement grpc.StatsHandler for trace instrumentation

### LINKED ISSUE #197: Implement `grpc.StatsHandler` for grpc instrumentation
See OpenCensus https://godoc.org/go.opencensus.io/plugin/ocgrpc

Much easier to use and implement, now:
```golang
s := grpc.NewServer(
		grpc.UnaryInterceptor(grpctrace.UnaryServerInterceptor(global.Tracer(""))),
		grpc.StreamInterceptor(grpctrace.StreamServerInterceptor(global.Tracer(""))),
	)
```
After:
```
s := grpc.NewServer(grpc.StatsHandler(&grpctrace.NewServerHandler(global.Tracer(""))))
```

### PR DESCRIPTION
**Pull request by fatsheep9146**
fixed #197
related #2840
Signed-off-by: Ziqi Zhao <zhaoziqi9146@gmail.com>

### CODE REVIEW

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go
Comment (hanyuancheung) [inline instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:28]: The import part should be divided into different partitions.

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go
Comment (hanyuancheung) [inline instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:31]:
```suggestion
type gRPCContext struct {
```
grpc is a Proper Noun. Rename it into gRPCContext.

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go
Comment (hanyuancheung) [inline instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:36]:
```suggestion
// NewServerHandler creates a stats.Handler for gRPC server.
```

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go
Comment (hanyuancheung) [inline instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:46]:
```suggestion
```
Meaningless blank line, should be removed.

**Comment by fatsheep9146** [comment]
@MrAlias @dashpole @dmathieu 
Please help review this pr, thanks :)

**Comment by dashpole** [comment]
Thanks for tackling this!

I have a few questions. Can you update the description with the answers?

* Is this meant to have feature parity with the existing otelgrpc instrumentation, and eventually replace it?  Or is it meant to be something you use instead of otelgrpc instrumentation?  If it is replacing it, what is the plan for deprecating the old one?
* Is there anything you were unable to implement with the stats handler interface?

One other question:

Is it possible to run the same unit tests against a stats-handler-based implementation and the current implementation?  Or did you copy existing test cases when writing these tests?

**Comment by fatsheep9146** [comment]
Sorry for my delay

> Is this meant to have feature parity with the existing otelgrpc instrumentation, and eventually replace it?  Or is it meant to be something you use instead of otelgrpc instrumentation?  If it is replacing it, what is the plan for deprecating the old one?

I think this feature is not to replace the existing one, but to provide a new way to instrument the grpc library. But in this new way with stats handler, we are able to implement more cases that the interceptors do not, for example, a failure to decode the message which is mentioned in https://github.com/open-telemetry/opentelemetry-go-contrib/issues/197#issuecomment-668377700.

> Is there anything you were unable to implement with the stats handler interface?

For now, I think stats handler has much more info about the grpc request, response and connection, so it definitely is a good place to add instrument logic. I can not answer this question based on what I know about the grpc stats handler.

> Is it possible to run the same unit tests against a stats-handler-based implementation and the current implementation? Or did you copy existing test cases when writing these tests?

Yes, I already add the same test like `interceptors` in test directory.

@dashpole

**Comment by fatsheep9146** [comment]
@open-telemetry/go-approvers hi, is anyone able to review this pr? Thanks a lot

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go
Comment (s-matyukevich) [inline instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:159]: I think we now can also set the following 2 [attributes](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/instrumentation/google.golang.org/grpc/otelgrpc/semconv.go#L34-L38), which, if I am reading the code right, are completely ignored. rs.Length and rs.WireLength properties should contain this info. 

Looks like those properties were deleted [here](https://github.com/open-telemetry/opentelemetry-go-contrib/pull/3168/files#diff-3218c106290c074c8331304ce04eb5006aa78fb8b913a2252e51930cfc7d3a1bL52) because caclulating them in interceptors was too expensive. With StatsHandler version we no longer have this problem.

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go
Comment (ash2k) [inline instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:70]: Here and blow: I think it'd be better to use a separate dedicated type for the key. Typically, an alias for `struct{}`. Benefits:
- It's confusing that this type is used for two purposes.
- As the key, it doesn't carry any data (it doesn't need to), but consumes 16 bytes of memory. This is pure waste that can be avoided by just using a separate type.

WDYT?

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go
Comment (fatsheep9146) [reply instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:161]: Done

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go
Comment (ash2k) [inline instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:61]: I think it'd be slightly more performant to invert the append operation:
- Current code allocates `attrs`, `mAttrs` and appends them into a new slice. I.e. 3 allocated slices in the worst case.
- `mAttrs = append(mAttrs, RPCSystemGRPC)` allocates `mAttrs` and appends them into a new slice. I.e. 2 allocated slices in the worst case.

Comment (fatsheep9146) [reply instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go]: Thanks! You are right, I modify this place and another place which has similar problem as you said.
@ash2k

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go
Comment (pellared) [inline instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:21]: why not simply?

```suggestion
	"google.golang.org/grpc/codes"
```

File: CHANGELOG.md
Comment (pellared) [inline CHANGELOG.md:14]: This is not a refactoring[^1] as it changes the output.
Can you describe the changes more precise?

[^1]: Refactoring is a disciplined technique for restructuring an existing body of code, altering its internal structure without changing its external behavior.

File: instrumentation/google.golang.org/grpc/otelgrpc/test/stats_handler_test.go
Comment (pellared) [inline instrumentation/google.golang.org/grpc/otelgrpc/test/stats_handler_test.go:33]: Why do we need this test? Is it not redunant with `TestStatsHandler`? `TestStatsHandler` also looks more "end to end" than this one.

**Review by pellared — COMMENTED** [review]
Leaving a quick review feedback

File: CHANGELOG.md
Comment (fatsheep9146) [reply CHANGELOG.md]: Done

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go
Comment (fatsheep9146) [reply instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:21]: Because I also import `go.opentelemetry.io/otel/codes` which has same package name.

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go
Comment (fatsheep9146) [reply instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go]: I use alias type `gRPCContextKey` as the key for gPRCContext, which is alias for `struct{}`
@ash2k

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go
Comment (ash2k) [inline instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:63]: May I suggest to implement options to configure this behavior. Similar to how `net/http` instrumentation works, see:

https://github.com/open-telemetry/opentelemetry-go-contrib/blob/75d0c13f6129330f3b5b060791aa8cefdc553849/instrumentation/net/http/otelhttp/config.go#L106-L124

I.e. default to private endpoints (i.e. `parent` relationship between spans/traces) and switch to public (i.e. `link` relationship) when explicitly asked to.

p.s. I think maintainers should probably decide and document how this should work across all integrations to provide consistent experience to users.

**Review by dashpole — COMMENTED** [review]
The code changes LGTM, but I think this needs to be surfaced in the README.  There, we should cover:

* Do we consider the new Stats Handlers experimental?
* Do we recommend users use interceptors, stats handlers, or both?
* What happens if users use both?  Will users get duplicate telemetry?  Will parent/child relationships between spans be broken because propagators are used in both interceptor and stats handlers instrumentation?  Or is the telemetry they get from stats handlers and interceptors complementary and non-overlapping?

Thanks for taking this on!

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go
Comment (ash2k) [reply instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:63]: It does maintain the existing behavior but at the same time now is the chance to introduce new default behavior, consistent with `net/http` instrumentation. If not now, then it'll be a breaking change for anyone who switches to the new implementation. Breaking change may be fine if this implementation is marked as experimental, I don't mind that or addressing it separately, I just need such an option. Thanks!

Comment (fatsheep9146) [reply instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:63]: > It does maintain the existing behavior but at the same time now is the chance to introduce new default behavior, consistent with `net/http` instrumentation. If not now, then it'll be a breaking change for anyone who switches to the new implementation. Breaking change may be fine if this implementation is marked as experimental, I don't mind that or addressing it separately, I just need such an option. Thanks!

I will open another issue to tackle this problem after this pr is merged.

**Comment by fatsheep9146** [comment]
@dashpole I start to try to answer the questions 

> Do we consider the new Stats Handlers experimental?

Yes, since the way of using this is a breaking change to user, so I think it is definitely experimental and need to be watched for a period of time to be stable.

> Do we recommend users use interceptors, stats handlers, or both?

I think we should recommend users to use stats handlers for two main reasons.

- **Functional advantages**: `stats` has more information for user to build more flexible and granular metric, for example
  - multiple different types of represent ”data length”: In [InPayLoad](https://pkg.go.dev/google.golang.org/grpc/stats#InPayload), there exists `Length`, `CompressedLength`, `WireLength` to denote the size of uncompressed, compressed payload data, with or without framing data. But in interceptors, we can only got uncompressed data, and this feature is also removed due to performance problem. https://github.com/open-telemetry/opentelemetry-go-contrib/pull/3168
  - more accurate timestamp: `InPayload.RecvTime` and `OutPayload.SentTime` records more accurate timestamp that server got and sent the message, the timestamp recorded by interceptors depends on the location of this interceptors in the total interceptor chain.
  - some other use cases: https://github.com/open-telemetry/opentelemetry-go-contrib/issues/197#issuecomment-668377700
 - **Performance advantages**: If too many interceptors are registered in a service, the interceptor chain can become too long, which increases the latency and processing time of the entire RPC call.

> What happens if users use both? Will users get duplicate telemetry? Will parent/child relationships between spans be broken because propagators are used in both interceptor and stats handlers instrumentation? Or is the telemetry they get from stats handlers and interceptors complementary and non-overlapping?
 
TBD

**Comment by fatsheep9146** [comment]
Could you help check this pr again? @dashpole

**Comment by dashpole** [comment]
Can you add your findings in https://github.com/open-telemetry/opentelemetry-go-contrib/pull/3002#issuecomment-1643278070 to a README?  I think it is important that users can understand the differences between the two implementations.

**Comment by fatsheep9146** [comment]
> Can you add your findings in [#3002 (comment)](https://github.com/open-telemetry/opentelemetry-go-contrib/pull/3002#issuecomment-1643278070) to a README? I think it is important that users can understand the differences between the two implementations.

Absolutely!

**Comment by fatsheep9146** [comment]
The README is already added, and I will add the example of using `stats.Handler` in [example](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrumentation/google.golang.org/grpc/otelgrpc/example) after this pr is merged.

@dashpole

**Comment by pellared** [comment]
nit: Can you please avoid doing force push as it makes reviewing harder? I cannot see the diff between reviews and the PR is quite large so it is easy to miss important issues in subsequent review integrations 😉

**Comment by fatsheep9146** [comment]
> nit: Can you please avoid force pushing as it makes reviewing harder? I cannot see the diff between reviews and the PR is quite large so it is easy to miss important issues in subsequent review integrations. Just make regular merge commits without rebasing 😉

Sorry about this. I usually prefer to clean up commit history in the end. I will notice this in the future. @pellared

**Comment by pellared** [comment]
> > nit: Can you please avoid force pushing as it makes reviewing harder? I cannot see the diff between reviews and the PR is quite large so it is easy to miss important issues in subsequent review integrations. Just make regular merge commits without rebasing 😉
> 
> Sorry about this. I usually prefer to clean up commit history in the end. I will notice this in the future. @pellared

> I usually prefer to clean up commit history in the end. 

We do "rebase and merge" so the git history is not a big issue for us.

PS. I am fine with cleaning up the git history - just avoid rebasing the commits that are already pushed 😄

**Comment by fatsheep9146** [comment]
@dashpole I think the failed test is flaky, I have tried the test in may local laptop for many times. I still can not reproduce this fail. Could you help rerun the failed check?

**Comment by fatsheep9146** [comment]
@open-telemetry/go-approvers could any approvers help review this pr again?

**Comment by pellared** [comment]
@hanyuancheung Can you please review (as codeowner)?

**Comment by fatsheep9146** [comment]
@hanyuancheung nicely ping for review

**Comment by fatsheep9146** [comment]
@hanyuancheung ping for review, thanks

File: instrumentation/google.golang.org/grpc/otelgrpc/README.md
Comment (ash2k) [inline instrumentation/google.golang.org/grpc/otelgrpc/README.md:15]:
```suggestion
- **Functional advantages**: `stats.Handler` has more information for user to build more flexible and granular metric, for example
```

**Review by ash2k — APPROVED** [review]
Really looking forward to using the new instrumentation.

**Comment by fatsheep9146** [comment]
@open-telemetry/go-approvers 
Is there any possibilities we could push this pr forward?

**Review by hanyuancheung — APPROVED** [review]
LGTM

**Comment by fatsheep9146** [comment]
The failed test is still flaky, I will solve this in another pr.

**Comment by fatsheep9146** [comment]
@hanyuancheung could you help rerun the failed test?

**Comment by fatsheep9146** [comment]
@pellared could you help merge this pr?

**Comment by pellared** [comment]
@fatsheep9146 I just returned from a long PTO. I will do my best to look at this before end of this week.

**Comment by fatsheep9146** [comment]
> @fatsheep9146 I just returned from a long PTO. I will do my best to look at this before end of this week.

Thanks!

File: instrumentation/google.golang.org/grpc/otelgrpc/README.md
Comment (pellared) [inline instrumentation/google.golang.org/grpc/otelgrpc/README.md:8]: 1. typo
2. what is it noted as "experimental"?

```suggestion
- by gRPC `stats.Handler`
```

**Review by pellared — APPROVED** [review]
I propose to:

1. Merge this PR
2. Create an issue for https://github.com/open-telemetry/opentelemetry-go-contrib/pull/3002/files#r1209293763
3. Deprecate the interceptors so that they are available in next release.
4. After the next release remove the interceptors. 

@open-telemetry/go-approvers please upvote/downvote if you agree or not. Feel free to leave any additional feedback.

File: instrumentation/google.golang.org/grpc/otelgrpc/README.md
Comment (fatsheep9146) [reply instrumentation/google.golang.org/grpc/otelgrpc/README.md:8]: > 2\. what is it noted as "experimental"?

I wrote "experimental" to express that the feature is now not stable enough, and should be used for a period of time to reach the stable stage.

Do we have some common conventions to express this?
@pellared

Comment (fatsheep9146) [reply instrumentation/google.golang.org/grpc/otelgrpc/README.md:8]: > It would be better if we put the docs in `doc.go` instead of `README.md`. More:

I will move to doc.go

Comment (pellared) [reply instrumentation/google.golang.org/grpc/otelgrpc/README.md:8]: > I wrote "experimental" to express that the feature is now not stable enough, and should be used for a period of time to reach the stable stage.

The version of the module is `v0.*` 😉

Comment (fatsheep9146) [reply instrumentation/google.golang.org/grpc/otelgrpc/README.md:8]: > conventions

Got it

Comment (fatsheep9146) [reply instrumentation/google.golang.org/grpc/otelgrpc/README.md]: I already moved readme to doc.go

Comment (pellared) [reply instrumentation/google.golang.org/grpc/otelgrpc/README.md]: Lint is failing (probably you need to format `doc.go`)

Comment (fatsheep9146) [reply instrumentation/google.golang.org/grpc/otelgrpc/README.md]: > Lint is failing (probably you need to format `doc.go`)

Done

File: instrumentation/google.golang.org/grpc/otelgrpc/doc.go
Comment (pellared) [inline instrumentation/google.golang.org/grpc/otelgrpc/doc.go:1]: These doc comments do not render correctly.

image


See here for more guidance how to use the syntax and test it: https://github.com/open-telemetry/opentelemetry-go/blob/main/CONTRIBUTING.md#documentation

File: instrumentation/google.golang.org/grpc/otelgrpc/doc.go
Comment (pellared) [inline instrumentation/google.golang.org/grpc/otelgrpc/doc.go:39]:
```suggestion
[stats.InPayload]: https://pkg.go.dev/google.golang.org/grpc/stats#OutPayload
[stats.OutPayload]: https://pkg.go.dev/google.golang.org/grpc/stats#InPayload
```

and also update the casing in the references

**Review by pellared — APPROVED** [review]

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go
Comment (pellared) [inline instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:27]: It should be updated to 
```suggestion
	semconv "go.opentelemetry.io/otel/semconv/v1.17.0"
```

**Review by pellared — CHANGES_REQUESTED** [review]
I noticed that the stats handler misses the functionality you added here: https://github.com/open-telemetry/opentelemetry-go-contrib/pull/2700. Any reasons for that?

**Comment by fatsheep9146** [comment]
> I noticed that the stats handler misses the functionality you added here: #2700. Any reasons for that?

I was plan to implement the stats.handler step by step, cover the trace first, and then cover the metrics. Which is also convenient for review. 

Do you think I should implement it in this pr or create another pr for metrics?

I'm open for both way. @pellared

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go
Comment (fatsheep9146) [reply instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:27]: OK

**Comment by pellared** [comment]
> > I noticed that the stats handler misses the functionality you added here: #2700. Any reasons for that?
> 
> I was plan to implement the stats.handler step by step, cover the trace first, and then cover the metrics. Which is also convenient for review.
> 
> Do you think I should implement it in this pr or create another pr for metrics?
> 
> I'm open for both way. @pellared

During the SIG meetings we discussed that we would like to have a feature-parity before we deprecate it as part of https://github.com/open-telemetry/opentelemetry-go-contrib/pull/3002#pullrequestreview-1613182957.

Still I think we can tackle it in a separate PR.

Are there any other features which are missing when comparing with the interceptors?

**Comment by fatsheep9146** [comment]
> > > I noticed that the stats handler misses the functionality you added here: #2700. Any reasons for that?
> > 
> > 
> > I was plan to implement the stats.handler step by step, cover the trace first, and then cover the metrics. Which is also convenient for review.
> > Do you think I should implement it in this pr or create another pr for metrics?
> > I'm open for both way. @pellared
> 
> During the SIG meetings we discussed that we would like to have a feature-parity before we deprecate interceptors as part of [#3002 (review)](https://github.com/open-telemetry/opentelemetry-go-contrib/pull/3002#pullrequestreview-1613182957).
> 
> Still I think we can tackle it in a separate PR.
> 
> Are there any other features which are missing when comparing with the interceptors?

Only metrics 

> > > I noticed that the stats handler misses the functionality you added here: #2700. Any reasons for that?
> > 
> > 
> > I was plan to implement the stats.handler step by step, cover the trace first, and then cover the metrics. Which is also convenient for review.
> > Do you think I should implement it in this pr or create another pr for metrics?
> > I'm open for both way. @pellared
> 
> During the SIG meetings we discussed that we would like to have a feature-parity before we deprecate interceptors as part of [#3002 (review)](https://github.com/open-telemetry/opentelemetry-go-contrib/pull/3002#pullrequestreview-1613182957).
> 
> Still I think we can tackle it in a separate PR.
> 
> Are there any other features which are missing when comparing with the interceptors?

I think metric is the only missing feature, since I wrote a simple e2e test `TestStatsHandler` just like interceptors to make sure the trace related function is complete.
@pellared

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go
Comment (fatsheep9146) [reply instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go]: Done @pellared all checks are passed

**Review by pellared — APPROVED** [review]

**Review by dmathieu — APPROVED** [review]

**Comment by fatsheep9146** [comment]
@pellared Thanks for creating the following issues!

**Comment by hanyuancheung** [comment]
It seems that this PR could be merged first, and then we could continue the following work.

**Comment by pellared** [comment]
@fatsheep9146 Merged 🎉 Thank you for your patience ❤️ You can follow-up with https://github.com/open-telemetry/opentelemetry-go-contrib/issues/4316 and https://github.com/open-telemetry/opentelemetry-go-contrib/issues/4317

**Comment by fatsheep9146** [comment]
> @fatsheep9146 Merged 🎉 Thank you for your patience ❤️ You can follow-up with #4316 and #4317

Thanks! I will solve the follow-up ASAP. @pellared

**Comment by mehdihadeli** [comment]
Hi,
When do you release new style of grpc otel? V0.44 doesn't have grpc metrics

**Comment by dmathieu** [comment]
0.45.0 was released a few hour ago.
https://github.com/open-telemetry/opentelemetry-go-contrib/releases/tag/v1.20.0

**Comment by pellared** [comment]
> 0.45.0 was released a few hour ago. https://github.com/open-telemetry/opentelemetry-go-contrib/releases/tag/v1.20.0

Yet metrics are not supported yet. See https://github.com/open-telemetry/opentelemetry-go-contrib/issues/4316

**Comment by pellared** [comment]
> 0.45.0 was released a few hour ago. https://github.com/open-telemetry/opentelemetry-go-contrib/releases/tag/v1.20.0

@mehdihadeli  Yet metrics are not supported yet. See https://github.com/open-telemetry/opentelemetry-go-contrib/issues/4316. We hope to implement it for the next release.

**Pull request merged**
